langchain/vectorstores/typesense.py

```diff
@@ -0,0 +1,311 @@
+        text_key: str,
+    ):
+        """Initialize with Typesense client."""
+        try:
+            import typesense
+        except ImportError:
+            raise ValueError(
+                "Could not import typesense python package. "
+                "Please install it with `pip install typesense`."
+            )
+        if not isinstance(typesense_client, typesense.Client):
+            raise ValueError(
+                f"typesense_client should be an instance of typesense.Client, "
+                f"got {type(typesense_client)}"
+            )
+        self.typesense_client = typesense_client
+        self.typesense_collection_name = typesense_collection_name
+        self._embedding_function = embedding_function
+        self._text_key = text_key
+
+    def add_texts(
+        self,
+        texts: Iterable[str],
+        metadatas: Optional[List[dict]] = None,
+        ids: Optional[List[str]] = None,
+        **kwargs: Any,
+    ) -> List[str]:
+        """Run more texts through the embeddings and add to the vectorstore.
+
+        Args:
+            texts: Iterable of strings to add to the vectorstore.
+            metadatas: Optional list of metadatas associated with the texts.
+            ids: Optional list of ids to associate with the texts.
+
+        Returns:
+            List of ids from adding the texts into the vectorstore.
+
+        """
+        import time
+
+        from typesense.exceptions import ObjectAlreadyExists, ObjectNotFound
+
+        # Embed and create the documents
+        docs = []
+        ids = ids or [str(uuid.uuid4()) for _ in texts]
+        for i, text in enumerate(texts):
+            embedding = self._embedding_function(text)
+            metadata = metadatas[i] if metadatas else {}
+            docs.append(
+                {"id": ids[i], "vec": embedding, "text": text, "metadata": metadata}
+            )
+
+        # upsert to Typesense
+        while True:
+            try:
+                self.typesense_client.collections[
+                    self.typesense_collection_name
+                ].documents.import_(docs, {"action": "upsert"})
+            except ObjectNotFound:
+                # Create the collection if it doesn't already exist
+                try:
+                    self.typesense_client.collections.create(
+                        {
+                            "name": self.typesense_collection_name,
+                            "fields": [
+                                {
+                                    "name": "vec",
+                                    "type": "float[]",
+                                    "num_dim": len(docs[0]["vec"]),
+                                },
+                                {"name": "text", "type": "string"},
+                                {"name": ".*", "type": "auto"},
+                            ],
+                        }
+                    )
+                except ObjectAlreadyExists:
```

Comment:
how could we get an ObjectAlreadyExists if we only go down this path after an ObjectNotFound error?

Reply:
Hmmm, good point! Added it due to force of habit!

Should be safe to remove the exception handling here.